TPC Decoder module and Channel Map Service

tomjunk opened the pull request:
This PR introduces a ported TPC decoder module taken from the one in SBNDQM, with a renaming of its analysis data product and the use of a new channel map service for assigning offline channel numbers based on the FEB crate, FEB, and channel within the FEB.  The channel map service reads a map file made from a spreadsheet produced by Nupur Oza.  A future feature planned is to add functionality to access the hardware database to  read map information, but for now, a file is used.  There are fcl configuration files for running the decoder.  The decoder is moved to sbndcode so it can run as part of the offline reco.

fjnicolas commented:
Hi @tomjunk, thanks for the PR! Can you add some reviewers? Thank you!

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/20460&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/12585&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/12585&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/20460&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/12586&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/12586&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/20461&builds=)

tomjunk commented:
Hi Francisco -- thanks much!  I don't think I have write access to this repo and thus can only suggest reviewers in a comment like this.  Daisy Kalra and Nupur Oza are the most closely connected people to this effort.  Gray Putnam wrote the decoder for sbndqm and this is just a port to sbndcode with a call to the new channel map, so he is a candidate to review as well.

I see there are some problems with the c7 build however (e20 looks fine).  Here's the log -- they look like issues with Clang not liking the Nevis TPC Fragment:

4833: In file included from /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndcode/sbndcode/Decoders/TPC/SBNDTPCDecoder_module.cc:11:
4834: In file included from /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndcode/sbndcode/Decoders/TPC/SBNDTPCDecoder.h:22:
4835: In file included from /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPCFragment.hh:5:
4836: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:24:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4837:     struct {
4838:     ^
4839: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:44:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4840:     struct {
4841:     ^
4842: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:58:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4843:     struct {
4844:     ^
4845: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:72:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4846:     struct {
4847:     ^
4848: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:86:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4849:     struct {
4850:     ^
4851: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndaq_artdaq_core/sbndaq-artdaq-core/Overlays/SBND/NevisTPC/NevisTPCTypes.hh:100:5: error: anonymous types declared in an anonymous union are an extension [-Werror,-Wnested-anon-types]
4852:     struct {
4853:     ^
4854: In file included from /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndcode/sbndcode/Decoders/TPC/SBNDTPCDecoder_module.cc:11:
4855: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndcode/sbndcode/Decoders/TPC/SBNDTPCDecoder.h:96:12: error: private field '_last_event_number' is not used [-Werror,-Wunused-private-field]
4856:   uint32_t _last_event_number;
4857:            ^
4858: /scratch/workspace/sbnd_ci/label_exp/SLF7/label_exp2/swarm/SBND/srcs/sbndcode/sbndcode/Decoders/TPC/SBNDTPCDecoder.h:97:12: error: private field '_last_trig_frame_number' is not used [-Werror,-Wunused-private-field]
4859:   uint32_t _last_trig_frame_number;
4860:            ^
4861: 8 errors generated.

tomjunk commented:
Actually two of these look like unused variables in the dcoder itself.  I'll get rid of those and try compiling with clang.

tomjunk commented:
Okay, I found the line in sbndaq-art-daq-core that turns off the warning about nested anonymous structs in unions.  It was there to make sbndaq-artdaq-core build, but programs that #include its headers also need that switch.  I also removed some unused variables and their initializers, and now clang is happy.  I updated the sbncode version to v09_78_02 from v09_78_00 in order to make my local build happy.

fjnicolas commented:
Thanks for quick fix for clang! I've asked you to be added to the SBN related repos. I'll add the suggested reviewers meanwhile.

kalradaisy reviewed (approved):
I approve these changes.

fjnicolas commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/20475&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/20476&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/12598&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/12598&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/20475&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/12599&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/12599&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/20476&builds=)
